### DISCUSSION TITLE
design tokens

### DISCUSSION THREAD

**Discussion by poyribn** [godotengine/godot-proposals#7429] [GUI]
'design tokens' are variables that affect a GUI theme overall, like _panel corner radius_ or _accent color_.

this would be useful for [DARS generator](https://sosasees.itch.io/dars-generator), an app that switches between 3 near-identical themes where only 3 colors are swapped out in each. in DARS generator version [2023].0, these themes are actual duplicated files.

design tokens would also be very useful in future games, where i want to change the GUI colors based on the environment where the player character is, and based on accessibility settings like 'high contrast'.

most likely it's much better to not build design tokens into Godot, but expect the developer to procedurally generate the theme with a script if it should adjust based on changing design tokens.
maybe there should be a repository of free and open-source scripts for procedurally generating themes, each one having a different visual style and different design tokens exposed as variables in the Inspector.

**Comment by theraot** [comment]
You don't need the extremes of duplicated themes or generating a complete theme from a script. Although both extremes are possible.

You can instead: modify the theme from a script. For the use case of only a few properties being different this should not be much trouble.

And if you are going to have a method that modifies the theme in multiple ways for different situations, it might have a parameter you use to tell which set of changes to make, and you might call that parameter "design token".

**Comment by Calinou** [comment]
> maybe there should be a repository of free and open-source scripts for procedurally generating themes, each one having a different visual style and different design tokens exposed as variables in the Inspector.

I started working on such a prototype, but didn't have time to finish it: https://github.com/Calinou/godot-theme-generator

It is definitely possible, it's just a fair amount of effort (especially if you want to provide enough flexibility for real world projects).

**Comment by inhaledesign** [comment]
Would like to voice support for design tokens. They make styling an application much more consistent.

> most likely it's much better to not build design tokens into Godot

I'm confused by this statement. I can't imagine adding support for design tokens would be a big lift. All you need is for the theme system to support variables, and for variables to be able to reference other variables. This is how tokens are done in Figma, e.g.

**Reply by poyribn** [reply]
I did think Design Tokens are much more complicated under the Hood.
Thank you for the Correction 🤍

**Comment by Calinou** [comment]
Thanks for the proposal! Consolidating in https://github.com/godotengine/godot-proposals/issues/13947.
